src/stream/src/executor/top_n/top_n_cache.rs

```diff
@@ -239,46 +255,34 @@ impl<const WITH_TIES: bool> TopNCache<WITH_TIES> {
         full
     }
 
-    pub fn is_high_cache_full(&self) -> bool {
+    pub fn high_is_full(&self) -> bool {
         // For WITH_TIES, the high cache can exceed the capacity.
         if !WITH_TIES {
-            assert!(self.high.len() <= self.high_capacity);
+            assert!(self.high.len() <= self.high_cache_capacity);
         }
-        self.high.len() >= self.high_capacity
+        self.high.len() >= self.high_cache_capacity
     }
 
-    fn last_cache_key_before_high(&self) -> Option<&CacheKey> {
-        let middle_last_key = self.middle.last_key_value().map(|(k, _)| k);
-        middle_last_key.or_else(|| self.low.last_key_value().map(|(k, _)| k))
-    }
-
-    /// Use this method instead of `self.high.insert` directly when possible.
-    ///
-    /// It only inserts into high cache if the key is smaller than the largest key in the high
-    /// cache. Otherwise, we simply ignore the row. We will wait until the high cache becomes
-    /// empty and fill it at that time.
-    fn insert_high_cache(&mut self, cache_key: CacheKey, row: CompactedRow, is_from_middle: bool) {
-        if !self.is_high_cache_full() {
-            if is_from_middle {
-                self.high.insert(cache_key, row);
-                return;
-            }
-            // For direct insert, we need to check if the key is smaller than the largest key
-            if let Some(high_last) = self.high.last_key_value()
-                && cache_key <= *high_last.0
-            {
-                debug_assert!(cache_key != *high_last.0, "cache_key should be unique");
-                self.high.insert(cache_key, row);
-            }
+    fn high_is_synced(&self) -> bool {
+        if !self.high.is_empty() {
+            true
         } else {
-            let high_last = self.high.last_entry().unwrap();
-            if cache_key <= *high_last.key() {
-                debug_assert!(cache_key != *high_last.key(), "cache_key should be unique");
-                high_last.remove_entry();
```

Comment:
This is a potential bug. When `WITH_TIES = true`, this will leave the high cache in a inconsistent state, not containing all ties of the last/largest element.

Reply:
Is it possible to have a test for the case?

Reply:
I mean, we did have some complex bug for TopN before https://github.com/risingwavelabs/risingwave/pull/8659.